bip-0077.mediawiki

```diff
@@ -0,0 +1,282 @@
+* The directory awaits a request from the sender if it goes offline. Upon request, the directory relays the encrypted <code>Payjoin PSBT</code>.
+* The sender validates the <code>Payjoin PSBT</code> according to [[#senders-payjoin-psbt-checklist|the sender checklist]], signs its inputs and broadcasts the transaction to the Bitcoin network.
+
+The Original PSBT MUST:
+
+* Include complete UTXO data.
+* Be signed.
+* Exclude unnecessary fields such as global xpubs or keypath information.
+* Be broadcastable.
+
+The Original PSBT MAY:
+
+* Include outputs unrelated to the sender-receiver transfer for batching purposes.
+
+The Payjoin PSBT MUST:
+
+* Include all inputs from the Original PSBT.
+* Include all outputs which do not belong to the receiver from the Original PSBT.
+* Include complete UTXO data.
+
+The Payjoin PSBT sender MAY:
+
+* Add, remove or modify Original PSBT outputs under the control of the receiver (i.e. not sender change).
+
+The Payjoin PSBT MUST NOT:
+
+* Shuffle the order of inputs or outputs; the additional outputs or additional inputs must be inserted at a random index.
+* Decrease the absolute fee of the Original PSBT.
+
+====Session Initiation====
+
+Receivers must initialize a Payjoin Session on a directory to have a Session subdirectory allocated to them, as follows:
+
+A receiver must first discover the directory's OHTTP Gateway Key Configuration via an authenticated bootstrap mechanism before it can initialize a Session to receive payjoin requests. This mechanism may vary by implementation but must follow [[https://datatracker.ietf.org/doc/html/draft-ietf-privacypass-key-consistency-01| OHTTP Consistency Requirements]] and should not reveal a receiver IP address to the directory. Some examples of suitable mechanisms include fetching over a VPN, using keys included with an application binary, https-in-http CONNECT method, https-in-WebSocket, or a Tor hidden service.
+
+Payjoin Sessions begin by having a receiver send the static public key to the directory via OHTTP, receiving their subdirectory in a base64url encoded directory URL as response. Initiation may be replayed in case the receiver goes offline and should result in the same response, i.e. it Is idempotent.
+
+Optionally, before returning the URI, the receiver may request an authentication token by presenting a message containing only the word <code>Authenticate: <description></code>, after which the receiver is required to submit an <code>Authenticate: <token></code> including the token from the directory out of band. If authentication fails, an error is returned.
+
+If a directory is operated by an exchange, it may give out authentication tokens for users of its app, or may require some proof of work out of band. Tokens should be anonymous credentials from the directory describing the parameters of their authorization. Specific credentialing is out of the scope of this proposal.
```

Comment:
> * shouldn't the directory indicate to the receiver that authorization is required using a 401 response?

Yes, that's proper HTTP semantics

> * wouldn't an `Authorization` header be more appropriate? Specifically RFC 6750? the wording "required to submit an `Authenticate: <token>`" doesn't indicate how it's to be submitted (seems to imply that it's as a header but closest is `WWW-Authenticate` which is a response header used in 40{1,7} responses)

Yes

> * instead of "should be anonymous credentials", if a token is used maybe it's better to specify that it "must be unlinkable"?

Yes

TODO